pages/send_email_page.py

```diff
@@ -0,0 +1,120 @@
+            combined_df.iloc[:, 2:] = combined_df.iloc[:, 2:].apply(pandas.to_numeric, errors='coerce')
+            grouped_df = combined_df.groupby([TestData.excel_row_degaline, TestData.excel_row_adresas]).first()
+
+            email_content = ''
+            for (gas_station_name, gas_station_address), row in grouped_df.iterrows():
+                email_content += f"Gas station: {gas_station_name} "
+                email_content += f" Address: {gas_station_address}\n"
+                for gas_type, price in row.items():
+                    if pandas.notna(price):
+                        email_content += f"{gas_type} price: {price} EUR\n"
+                email_content += "\n"
+            return email_content
+        except Exception as e:
+            logger.error(f"Error occurred creating email content: {str(e)}")
+
+    def extract_data_to_json(self):
+        self.extract_data_from_excel_to_json(TestData.file_path_excel, TestData.file_path_json, self.logger)
+
+    def open_gmail(self):
+        try:
+            self.logger.info(f"Initializing new window")
+            self.open_new_window()
+            self.switch_to_new_window()
+            self.get_url(TestData.url_gmail)
+            self.logger.info(f"New window opened GMAIL main page loaded")
+        except Exception as e:
+            self.logger.error(f"Open GMAIL in new window failed: {str(e)}")
+
+    def login_to_gmail(self):
+        try:
+            self.logger.info(f"Start to login to GMAIL")
+            self.send_keys(GMAIL_EMAIL_INPUT, TestData.gmail_sign_in_email)
+            self.click(GMAIL_NEXT_BUTTON)
+            self.wait_element(GMAIL_PASSWORD_INPUT)
+            self.send_keys(GMAIL_PASSWORD_INPUT, TestData.gmail_sign_in_password)
+            self.click(GMAIL_NEXT_BUTTON)
+            self.logger.info(f"Logged in to GMAIL")
+        except Exception as e:
+            self.logger.error(f"Login to gmail failed: {str(e)}")
+
+    def create_new_mail_insert_recipient(self, email=None):
+        try:
+            self.logger.info(f"Start to create new EMAIL")
+            self.wait_element(GMAIL_NEW_MAIL)
+            self.click(GMAIL_NEW_MAIL)
+            self.wait_element(GMAIL_NEW_MAIL_TO)
+            recipient_email = email or self.default_recipient
+            self.send_keys(GMAIL_NEW_MAIL_TO, recipient_email)
+            self.logger.info(f"EMAIL created, recipient added")
+        except Exception as e:
+            self.logger.error(f"Create new email failed: {str(e)}")
+
+    def insert_subject_and_body(self):
+        try:
+            self.wait_element(GMAIL_NEW_MAIL_SUBJECT)
+            self.send_keys(GMAIL_NEW_MAIL_SUBJECT, TestData.gmail_subject)
+            self.wait_element(GMAIL_NEW_MAIL_INPUT)
+            email_content = self.create_email_content(TestData.file_path_json, self.logger)
+            self.send_keys(GMAIL_NEW_MAIL_INPUT, email_content)
+            self.logger.info(f"Subject added, email, content added")
```

Comment:
cia taip pat kaip ir visuose, stringuose, not optimal, kur nereikia, pasitikrink kur naudoji formated stringa jei nesiruosi passint jokio value {} viduje ar kitu budu, nutrink f pries stringo pradzia